File: medbrains/crates/medbrains-core/src/housekeeping.rs
use chrono::{DateTime, NaiveDate, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

// ── Enums ──────────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "cleaning_area_type", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum CleaningAreaType {
    Icu,
    Ward,
    Ot,
    Er,
    Lab,
    Pharmacy,
    Corridor,
    Lobby,
    Washroom,
    Kitchen,
    General,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "cleaning_task_status", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum CleaningTaskStatus {
    Pending,
    Assigned,
    InProgress,
    Completed,
    Verified,
    Rejected,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "linen_status", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum LinenStatus {
    Clean,
    InUse,
    Soiled,
    Washing,
    Condemned,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "linen_contamination_type", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum LinenContaminationType {
    Regular,
    Contaminated,
    Isolation,
}

// ── Structs ────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct CleaningSchedule {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub area_type: CleaningAreaType,
    pub location_id: Option<Uuid>,
    pub department_id: Option<Uuid>,
    pub frequency_hours: i32,
    pub checklist_items: serde_json::Value,
    pub is_active: bool,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct CleaningTask {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub schedule_id: Option<Uuid>,
    pub location_id: Option<Uuid>,
    pub department_id: Option<Uuid>,
    pub area_type: CleaningAreaType,
    pub task_date: NaiveDate,
    pub assigned_to: Option<String>,
    pub status: CleaningTaskStatus,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub verified_by: Option<Uuid>,
    pub verified_at: Option<DateTime<Utc>>,
    pub checklist_results: serde_json::Value,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct RoomTurnaround {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub location_id: Option<Uuid>,
    pub patient_id: Option<Uuid>,
    pub discharge_at: Option<DateTime<Utc>>,
    pub dirty_at: Option<DateTime<Utc>>,
    pub cleaning_started_at: Option<DateTime<Utc>>,
    pub cleaning_completed_at: Option<DateTime<Utc>>,
    pub ready_at: Option<DateTime<Utc>>,
    pub turnaround_minutes: Option<i32>,
    pub cleaned_by: Option<String>,
    pub verified_by: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct PestControlSchedule {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub location_id: Option<Uuid>,
    pub department_id: Option<Uuid>,
    pub pest_type: String,
    pub frequency_months: i32,
    pub last_done: Option<NaiveDate>,
    pub next_due: Option<NaiveDate>,
    pub vendor_name: Option<String>,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct PestControlLog {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub schedule_id: Option<Uuid>,
    pub treatment_date: NaiveDate,
    pub treatment_type: String,
    pub chemicals_used: Option<String>,
    pub areas_treated: serde_json::Value,
    pub vendor_name: Option<String>,
    pub certificate_no: Option<String>,
    pub next_due: Option<NaiveDate>,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct LinenItem {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub barcode: Option<String>,
    pub item_type: String,
    pub current_status: LinenStatus,
    pub ward_id: Option<Uuid>,
    pub wash_count: i32,
    pub max_washes: i32,
    pub commissioned_date: Option<NaiveDate>,
    pub condemned_date: Option<NaiveDate>,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct LinenMovement {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub linen_item_id: Option<Uuid>,
    pub movement_type: String,
    pub from_ward: Option<Uuid>,
    pub to_ward: Option<Uuid>,
    pub quantity: i32,
    pub weight_kg: Option<rust_decimal::Decimal>,
    pub contamination_type: LinenContaminationType,
    pub batch_id: Option<Uuid>,
    pub recorded_by: Option<String>,
    pub movement_date: DateTime<Utc>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct LaundryBatch {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub batch_number: String,
    pub items_count: i32,
    pub total_weight: Option<rust_decimal::Decimal>,
    pub contamination_type: LinenContaminationType,
    pub wash_formula: Option<String>,
    pub wash_temperature: Option<i32>,
    pub cycle_minutes: Option<i32>,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub status: String,
    pub operator_name: Option<String>,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct LinenParLevel {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub ward_id: Option<Uuid>,
    pub item_type: String,
    pub par_level: i32,
    pub current_stock: i32,
    pub reorder_level: i32,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct LinenCondemnation {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub linen_item_id: Option<Uuid>,
    pub reason: String,
    pub wash_count_at_condemn: Option<i32>,
    pub condemned_by: Option<Uuid>,
    pub condemned_date: NaiveDate,
    pub replacement_requested: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

File: medbrains/packages/types/src/housekeeping.ts
// Housekeeping types — split from index.ts, barrel-re-exported.

// ── Housekeeping ──────────────────────────────────────────

export type CleaningAreaType =
  | "icu"
  | "ward"
  | "ot"
  | "er"
  | "lab"
  | "pharmacy"
  | "corridor"
  | "lobby"
  | "washroom"
  | "kitchen"
  | "general";
export type CleaningTaskStatusType =
  | "pending"
  | "assigned"
  | "in_progress"
  | "completed"
  | "verified"
  | "rejected";
export type LinenStatusType = "clean" | "in_use" | "soiled" | "washing" | "condemned";
export type LinenContaminationTypeValue = "regular" | "contaminated" | "isolation";

export interface CleaningSchedule {
  id: string;
  tenant_id: string;
  area_type: CleaningAreaType;
  location_id?: string;
  department_id?: string;
  frequency_hours: number;
  checklist_items: unknown;
  is_active: boolean;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface CleaningTask {
  id: string;
  tenant_id: string;
  schedule_id?: string;
  location_id?: string;
  department_id?: string;
  area_type: CleaningAreaType;
  task_date: string;
  assigned_to?: string;
  status: CleaningTaskStatusType;
  started_at?: string;
  completed_at?: string;
  verified_by?: string;
  verified_at?: string;
  checklist_results: unknown;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface RoomTurnaround {
  id: string;
  tenant_id: string;
  location_id?: string;
  patient_id?: string;
  discharge_at?: string;
  dirty_at?: string;
  cleaning_started_at?: string;
  cleaning_completed_at?: string;
  ready_at?: string;
  turnaround_minutes?: number;
  cleaned_by?: string;
  verified_by?: string;
  created_at: string;
  updated_at: string;
}

export interface PestControlSchedule {
  id: string;
  tenant_id: string;
  location_id?: string;
  department_id?: string;
  pest_type: string;
  frequency_months: number;
  last_done?: string;
  next_due?: string;
  vendor_name?: string;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface PestControlLog {
  id: string;
  tenant_id: string;
  schedule_id?: string;
  treatment_date: string;
  treatment_type: string;
  chemicals_used?: string;
  areas_treated: unknown;
  vendor_name?: string;
  certificate_no?: string;
  next_due?: string;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface LinenItem {
  id: string;
  tenant_id: string;
  barcode?: string;
  item_type: string;
  current_status: LinenStatusType;
  ward_id?: string;
  wash_count: number;
  max_washes: number;
  commissioned_date?: string;
  condemned_date?: string;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface LinenMovement {
  id: string;
  tenant_id: string;
  linen_item_id?: string;
  movement_type: string;
  from_ward?: string;
  to_ward?: string;
  quantity: number;
  weight_kg?: number;
  contamination_type: LinenContaminationTypeValue;
  batch_id?: string;
  recorded_by?: string;
  movement_date: string;
  created_at: string;
  updated_at: string;
}

export interface LaundryBatch {
  id: string;
  tenant_id: string;
  batch_number: string;
  items_count: number;
  total_weight?: number;
  contamination_type: LinenContaminationTypeValue;
  wash_formula?: string;
  wash_temperature?: number;
  cycle_minutes?: number;
  started_at?: string;
  completed_at?: string;
  status: string;
  operator_name?: string;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface LinenParLevel {
  id: string;
  tenant_id: string;
  ward_id?: string;
  item_type: string;
  par_level: number;
  current_stock: number;
  reorder_level: number;
  created_at: string;
  updated_at: string;
}

export interface LinenCondemnation {
  id: string;
  tenant_id: string;
  linen_item_id?: string;
  reason: string;
  wash_count_at_condemn?: number;
  condemned_by?: string;
  condemned_date: string;
  replacement_requested: boolean;
  created_at: string;
  updated_at: string;
}

// Request types
export interface CreateCleaningScheduleRequest {
  area_type: string;
  location_id?: string;
  department_id?: string;
  frequency_hours?: number;
  checklist_items?: unknown;
  is_active?: boolean;
  notes?: string;
}
export interface UpdateCleaningScheduleRequest {
  area_type?: string;
  location_id?: string;
  department_id?: string;
  frequency_hours?: number;
  checklist_items?: unknown;
  is_active?: boolean;
  notes?: string;
}
export interface CreateCleaningTaskRequest {
  schedule_id?: string;
  location_id?: string;
  department_id?: string;
  area_type: string;
  task_date?: string;
  assigned_to?: string;
  notes?: string;
}
export interface UpdateTaskStatusRequest {
  status: string;
}
export interface CreateTurnaroundRequest {
  location_id?: string;
  patient_id?: string;
  discharge_at?: string;
  dirty_at?: string;
  cleaned_by?: string;
}
export interface CreatePestControlScheduleRequest {
  location_id?: string;
  department_id?: string;
  pest_type: string;
  frequency_months?: number;
  last_done?: string;
  next_due?: string;
  vendor_name?: string;
  notes?: string;
}
export interface UpdatePestControlScheduleRequest {
  pest_type?: string;
  frequency_months?: number;
  last_done?: string;
  next_due?: string;
  vendor_name?: string;
  notes?: string;
}
export interface CreatePestControlLogRequest {
  schedule_id?: string;
  treatment_date: string;
  treatment_type: string;
  chemicals_used?: string;
  areas_treated?: unknown;
  vendor_name?: string;
  certificate_no?: string;
  next_due?: string;
  notes?: string;
}
export interface CreateLinenItemRequest {
  barcode?: string;
  item_type: string;
  current_status?: string;
  ward_id?: string;
  max_washes?: number;
  commissioned_date?: string;
  notes?: string;
}
export interface UpdateLinenItemRequest {
  current_status?: string;
  ward_id?: string;
  notes?: string;
}
export interface CreateLinenMovementRequest {
  linen_item_id?: string;
  movement_type: string;
  from_ward?: string;
  to_ward?: string;
  quantity?: number;
  weight_kg?: number;
  contamination_type?: string;
  batch_id?: string;
  recorded_by?: string;
}
export interface CreateLaundryBatchRequest {
  batch_number: string;
  items_count?: number;
  total_weight?: number;
  contamination_type?: string;
  wash_formula?: string;
  wash_temperature?: number;
  cycle_minutes?: number;
  operator_name?: string;
  notes?: string;
}
export interface UpsertParLevelRequest {
  ward_id?: string;
  item_type: string;
  par_level: number;
  current_stock?: number;
  reorder_level?: number;
}
export interface CreateLinenCondemnationRequest {
  linen_item_id?: string;
  reason: string;
  wash_count_at_condemn?: number;
  replacement_requested?: boolean;
}
